classes/services/PKPFileService.inc.php

```diff
@@ -0,0 +1,251 @@
+	 */
+	public function __construct() {
+
+		$adapter = new Local(
+			Config::getVar('files', 'files_dir'),
+			LOCK_EX,
+			Local::DISALLOW_LINKS,
+			[
+				'file' => [
+					'public' => FILE_MODE_MASK,
+					'private' => FILE_MODE_MASK,
+				],
+				'dir' => [
+					'public' => DIRECTORY_MODE_MASK,
+					'private' => DIRECTORY_MODE_MASK,
+				]
+			]
+		);
+
+		HookRegistry::call('File::adapter', [$adapter, $this]);
+
+		$this->fs = new Filesystem($adapter);
+	}
+
+	/**
+	 * Get a file path by its id
+	 *
+	 * @param int $id
+	 * @return string
+	 */
+	public function getPath($id) {
+		$file = Capsule::table('files')
+			->where('file_id', '=', $id)
+			->first();
+		return $file ? $file->path : '';
+	}
+
+	/**
+	 * Add a file
+	 *
+	 * @param string $from
+	 * @param string $to
+	 * @return int file id
+	 */
+	public function add($from, $to) {
+		$stream = fopen($from, 'r+');
+		if (!$stream) {
+			throw new Exception("Unable to copy $from to $to.");
+		}
+		if (!$this->fs->writeStream($to, $stream)) {
+			throw new Exception("Unable to write file at $to.");
+		}
+		if (is_resource($stream)) {
+			fclose($stream);
+		}
+		return Capsule::table('files')->insertGetId(['path' => $to]);
+	}
+
+	/**
+	 * Delete an uploaded file
+	 *
+	 * @param int $id
+	 * @return File
+	 */
+	public function delete($id) {
+		$path = $this->getPath($id);
+		if (!$path) {
+			throw new Exception("Unable to locate file $id.");
+		}
+		if (!$this->fs->delete($path)) {
+			throw new Exception("Unable to delete file $id at $path.");
+		}
+		Capsule::table('files')
+			->where('file_id', '=', $id)
+			->delete();
+	}
+
+	/**
+	 * Download a file
+	 *
+	 * This method sends a HTTP response and ends the request handling.
+	 * No code will run after this method is called.
+	 *
+	 * @param int|string $idOrPath The id or path to the file
+	 * @param string $filename Filename to give to the downloaded file
+	 * @param boolean $inline Whether to stream the file to the browser
+	 */
+	public function download($idOrPath, $filename, $inline = false) {
+
+		if (is_int($idOrPath)) {
+			$path = $this->getPath($idOrPath);
+		} else {
+			$path = $idOrPath;
+		}
+
+		if (!$this->fs->has($path)) {
+			throw new Exception('File download failed because no file was found at ' . $path);
+		}
+
+		if (HookRegistry::call('File::download', [$idOrPath, &$filename, $inline])) {
+			return;
+		}
+
+		// Stream the file to the end user.
+		$localPath = rtrim(Config::getVar('files', 'files_dir'), '/') . '/' . $path;
+		$mimetype = $this->fs->getMimetype($path) ?? 'application/octet-stream';
+		$filesize = $this->fs->getSize($path);
+		header("Content-Type: $mimetype");
+		header("Content-Length: $filesize");
+		header('Accept-Ranges: none');
+		header('Content-Disposition: ' . ($inline ? 'inline' : 'attachment') . "; filename=\"$filename\"");
+		header('Cache-Control: private'); // Workarounds for IE weirdness
+		header('Pragma: public');
+
+		if (is_readable($localPath)) {
```

Comment:
This is pretty old code -- it would be nice to replace it with something more Flysystemy. (https://stackoverflow.com/questions/29266974/laravel-5-flysystem-download-file-from-remote-disk)

Reply:
Yes, I agree! I found a [hook](https://github.com/pkp/ojs/pull/2887/files#diff-77c9bbc41ea0245a26913e904e8edb3fb50686b505c076d39efbaba9762f0806R413) that fires _after_ the file has been streamed but before the request is finished. It seems to be related to the usage stats. So I think I just didn't want to touch the code too much in case i disturbed that, but if you have a recommendation for doing the file download _without_ ending the request that would be good.

(Or if you know why that hook is there and if/how we can safely remove it, that would be good too! It seems like the download method should end the request, but maybe there are good reasons not to bake that into the download method.)

Reply:
Yes, as far as I know it's only used for usage stats, and I don't think it's crucial there whether it fires before or after the download. It's possible that placing it at the end of the download is an attempt to catch only uninterrupted (completed) downloads. @ctgraham may know. It might be possible to trigger the hook after download using `register_shutdown_function` instead, if the streamy toolset doesn't allow another way.

Reply:
Yes, `FileManager::downloadFileFinished` is almost certainly used by the usage stats framework.  The `COUNTER` specification doesn't make demands on exactly how we identify downloads, just that we make a best effort to identify when someone has really downloaded the article for viewing purposes.  Broken/aborted downloads are probably a bit of an edge case, so if we don't have a clear technological way to consistently fire a hook at completion, consistently firing a hook at the start of the download would fit the spec, IMO.

Reply:
Thanks Clinton! Alec, the example you linked uses Laravel's `Response` object to stream the file. Flysystem is just getting the stream. I'm out of my depth here, but does this do what you were hoping? It seems to work on my end.

```diff
diff --git a/classes/services/PKPFileService.inc.php b/classes/services/PKPFileService.inc.php
index 263ec1d34..990380eec 100644
--- a/classes/services/PKPFileService.inc.php
+++ b/classes/services/PKPFileService.inc.php
@@ -132,7 +132,6 @@ class PKPFileService {
 		}
 
 		// Stream the file to the end user.
-		$localPath = rtrim(Config::getVar('files', 'files_dir'), '/') . '/' . $path;
 		$mimetype = $this->fs->getMimetype($path) ?? 'application/octet-stream';
 		$filesize = $this->fs->getSize($path);
 		header("Content-Type: $mimetype");
@@ -142,14 +141,8 @@ class PKPFileService {
 		header('Cache-Control: private'); // Workarounds for IE weirdness
 		header('Pragma: public');
 
-		if (is_readable($localPath)) {
-			$f = fopen($localPath, 'rb');
-			if (!$f) return false;
-			while (!feof($f)) {
-				echo fread($f, 4096);
-			}
-			fclose($f);
-		}
+		fpassthru($this->fs->readStream($path));
+		exit();
 	}
 
 	/**
```

Reply:
I have no idea if the complaints are still current, but apparently `fpassthru` has been memory-intensive in the past (see https://www.php.net/manual/en/function.fpassthru.php). Maybe let's just let sleeping dogs lie until we have a reason to change the implementation.

Reply:
(I just thought Flysystem might have an implementation that's trustworthy but I don't see one!)